src/group_theory/dihedral.lean

```diff
@@ -0,0 +1,394 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+We define finite dihedral groups, with elements denoted by
+`r i` and `s i` for `i : zmod n`.  We show that a pair of elements
+`r0, s0 ∈ G` satisfying appropriate relations gives rise to a
+homomorphism `Dₙ → G`, sending `r i` to `r0 ^ i` and `s i` to
+`(r0 ^ i) * s0`.  (We have chosen to formulate this in
+a way that works when `G` is merely a monoid, as that makes
+it easier to deal with `Dₙ`-actions as homomorphisms from
+`Dₙ` to endomorphism monoids.  Thus, our relations are
+`r0 ^ n = s0 * s0 = 1` and `r0 * s0 * r0 = s0`.)
+
+We also do the case n = ∞ separately.
+-/
+
+import data.fintype algebra.power_mod group_theory.cyclic
+
+namespace group_theory
+
+variable n : ℕ+
+
+@[derive decidable_eq]
+inductive dihedral
+| r : (zmod n) → dihedral
+| s : (zmod n) → dihedral
+
+namespace dihedral
+
+variable {n}
+
+def log : dihedral n → (zmod n)
+| (r i) := i
+| (s i) := i
+
+def is_s : dihedral n → bool
+| (r _) := ff
+| (s _) := tt
+
+def to_bz : dihedral n → bool × (zmod n)
+| (r i) := ⟨ff,i⟩
+| (s i) := ⟨tt,i⟩
+
+def of_bz : bool × (zmod n) → dihedral n
+| ⟨ff,i⟩ := r i
+| ⟨tt,i⟩ := s i
+
+variable (n)
+def bz_equiv : dihedral n ≃ (bool × (zmod n)) :=
+{ to_fun := to_bz,
+  inv_fun := of_bz,
+  left_inv := λ g, by {cases g; refl},
+  right_inv := λ bi, by {rcases bi with ⟨_|_,i⟩ ; refl} }
+
+def inc : cyclic n → dihedral n
+| (cyclic.r i) := r i
+
+variable {n}
+
+instance : fintype (dihedral n) :=
+fintype.of_equiv (bool × (zmod n)) (bz_equiv n).symm
+
+lemma card : fintype.card (dihedral n) = 2 * n :=
+calc
+  fintype.card (dihedral n) = fintype.card (bool × (zmod n)) :
+    begin rw[fintype.card_congr (bz_equiv n)], end
+  ... = (fintype.card bool) * (fintype.card (zmod n)) : by rw[fintype.card_prod]
+  ... = (fintype.card bool) * (fintype.card (fin n)) : rfl
+  ... = 2 * n : by rw[fintype.card_bool, fintype.card_fin]
```

Comment:
`by simp` should accomplish the last 3 steps

Reply:
The whole proof can be done by `by simp [fintype.card_congr (bz_equiv n), zmod.card_zmod]`.  Should this and `zmod.card_zmod` be marked `@[simp]`?  Should `zmod.card_zmod` be called `zmod.card`, or should this be called `dihedral.card_dihedral`?